src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/FlowCacheService.java

```diff
@@ -53,19 +60,32 @@ public FlowCacheService(PersistenceManager persistenceManager, Clock clock,
         flowStates = new HashMap<>();
         active = false;
         flowRepository = persistenceManager.getRepositoryFactory().createFlowRepository();
+        haSubFlowRepository = persistenceManager.getRepositoryFactory().createHaSubFlowRepository();
     }
 
-    private void initCache(FlowRepository flowRepository) {
-        for (Flow entry : flowRepository.findAll()) {
-            if (entry.isOneSwitchFlow()) {
+    private void initCache() {
+        for (Flow flow : flowRepository.findAll()) {
+            if (flow.isOneSwitchFlow()) {
                 continue;
             }
-            if (isIncompleteFlow(entry)) {
+            if (isIncompleteFlow(flow)) {
                 log.warn("Flow is incomplete, do not put it into flow cache (flow_id: {}, ctime: {}, mtime: {}",
-                        entry.getFlowId(), entry.getTimeCreate(), entry.getTimeModify());
+                        flow.getFlowId(), flow.getTimeCreate(), flow.getTimeModify());
                 continue;
             }
-            flowStates.put(entry.getFlowId(), FlowMapper.INSTANCE.toFlowState(entry));
+            flowStates.put(flow.getFlowId(), FlowMapper.INSTANCE.toFlowState(flow));
+        }
+
+        for (HaSubFlow haSubFlow : haSubFlowRepository.findAll()) {
+            if (haSubFlow.isOneSwitch()) {
+                continue;
+            }
+            if (isIncompleteHaFlow(haSubFlow.getHaFlow())) {
+                log.warn("HA-flow is incomplete, do not put it into HA-flow cache (flow_id: {}, ctime: {}, mtime: {}",
+                        haSubFlow.getHaFlowId(), haSubFlow.getTimeCreate(), haSubFlow.getTimeModify());
```

Comment:
I am not sure if those logs are important. If not I would remove them and use a stream().filter() function. If we can't remove, at least you can filter if is one switch

Reply:
Sorry, I explain it bad. I was suggesting using the forEach function instead.
Here is an example for flowRepository and flows.

```java
        flowRepository.findAll().stream()
                .filter(flow -> !flow.isOneSwitchFlow())
                .forEach(flow -> {
                    if (isIncompleteFlow(flow)) {
                        log.warn("Flow is incomplete, do not put it into flow cache (flow_id: {}, ctime: {}, mtime: {}",
                                flow.getFlowId(), flow.getTimeCreate(), flow.getTimeModify());
                        return;
                    }
                    flowStates.put(flow.getFlowId(), FlowMapper.INSTANCE.toFlowState(flow));
                });
```

And if the log wasn't obligatory it could be like this:

```java
        flowStates.putAll(flowRepository.findAll().stream()
                .filter(flow -> !flow.isOneSwitchFlow())
                .filter(flow -> !isIncompleteFlow(flow))
                .collect(Collectors.toMap(Flow::getFlowId, FlowMapper.INSTANCE::toFlowState)));
```

If you like the last example also the log could be delegated to the isIncomplete function.

BTW: if you want to negate the filter method you can use also the [Predicate.not() ](https://www.baeldung.com/java-negate-predicate-method-reference#predicate-not) method

Reply:
Updated. Now it looks much better. Thank you!

Reply:
much better yes!, thanks